neps/nep-0481.md

```diff
@@ -0,0 +1,145 @@
+---
+NEP: 481
+Title: Synchronous wasm submodules
+Authors: Moritz Zielke <moritz.zielke@aurora.dev>, Michael Birch <michael.birch@aurora.dev>
+Status: New
+DiscussionsTo: https://github.com/nearprotocol/neps/pull/0481
+Type: Protocol
+Version: 1.0.0
+Created: 2023-05-12
+LastUpdated: 2023-05-12
+---
+
+# Synchronous wasm submodules
+
+## Summary
+
+This NEP proposes to enable the synchronous execution of wasm submodules in the NEAR runtime. Submodules are deployed under a NEAR account (the _parent_). A submodule may contain arbitrary wasm instructions and has access to submodule host functions. The execution of a submodule is triggered by the parent contract invoking a host function to start the submodule. The submodule may suspend itself by yielding back to the parent, which may subsequently resume the submodule. Upon yield and resume, data can be exchanged bidirectionally between parent and submodule.
+
+## Motivation
+
+Currently all interactions between NEAR contracts are asynchronous, which prevents use cases that require synchronous interactions. An example of such a use case is Aurora.
+
+Aurora is a smart contract on NEAR which emulates an Ethereum ecosystem. To execute a function call on an Ethereum contract, it loads the contract and its state from storage, spins up an EVM interpreter to execute the contract’s EVM bytecode and persists state changes. Interpreting EVM bytecode inside a NEAR transaction is inefficient in terms of gas usage.
+
+Given a [compiler](https://github.com/aurora-is-near/evm2near) to convert EVM bytecode into wasm bytecode targeted at NEAR, Aurora can [reduce its gas consumption](https://github.com/aurora-is-near/aurora-engine/pull/463) significantly. Instead of spinning up an EVM interpreter and interpreting a contract’s EVM bytecode it would invoke a submodule whose wasm is equivalent that EVM bytecode.
+
+Significantly increasing the efficiency of Aurora can be benefical for the NEAR ecosystem in general. Beyond that, the modifications of NEAR that are required for synchronous wasm submodules can be a building block for [account extensions](https://gov.near.org/t/proposal-account-extensions-contract-namespaces/34227).
+
+### wasm submodules and account extensions
+
+The proposed wasm submodules can be interpreted as account extensions with tight restrictions. A wasm submodule is an account extension that can be executed only by the parent contract itself. It can access only submodule host functions which  allow it to yield back to the parent and exchange data with it. Host functions available to regular contracts cannot be invoked by the submodule. A submodule cannot read or write the parent’s storage and has no storage associated with itself.
+
+Due to its complexity, account extension functionality might be implemented in multiple stages. Synchronous wasm submodules could be an initial stage, which progresses towards account extensions as the restrictions mentioned above are lifted.
```

Comment:
I would rather look at these two different proposals not as two features that *might* potentially compose some day, but rather as proposals that *must* compose. I think keeping this in mind we will end up with something that's more generally useful and might even find new uses in future improvements to the protocol.

In fact, I believe there is very little that needs to be done to make sure this holds true. Today the two sets of proposals seem quite analogous to me in their attempt to introduce two distinct (OOP) objects with their own slightly differing semantics. If we instead separate data and methods we might quite easily end up in a place where we don't even need to draw any comparisons between the NEPs, because each one is useful on their own. In particular one way forward I could see is focusing entirely on introducing a mechanism to make a synchronous function call to an isolated wasm-core-1 module with a communication channel of some sort. Even without changes to the data model this feature can be used in some way by the contract making self-calls (and for Aurora specifically, they could temporarily bake contracts into their contract while the data model changes are underway; this would likely make this proposal much more straightforward to think 'bout & implement too)

Reply:
> If we instead separate data and methods we might quite easily end up in a place where we don’t even need to draw any comparisons between the NEPs, because each one is useful on their own.

Intuitively I’m thinking changes to the data model such as seperate state are useful only once there exist some kind of separate entities (like submodules or contracts deployed to a namespace) which can be called. Which makes me wonder how changes to the data model along the lines of NEP-480 could be useful on their own? Probably I’m missing or misunderstanding something here.

Reply:
When I was saying "data model changes" I meant specifically extending the protocol with a concept of an account extension or a submodule as described here. But I also think you misunderstood the order of steps I was suggesting to be taken. My thinking was that the data model changes would be the very last step once the actions and APIs necessary for both this and account extension are in place. But this is just my brainstorming. Ultimately the substance of the earlier message is:

> I would rather look at these two different proposals not as two features that *might* potentially compose some day, but as proposals that *must* compose. I think keeping this in mind we will end up with something that's more generally useful and might even find new uses in future improvements to the protocol.

I don't have a particularly strong opinion on how this is achieved, but I think it is important that we don't end up with both submodules *and* account extensions in the protocol, because unifying them is likely to be infeasible.